mesonbuild/cmake/interpreter.py

```diff
@@ -275,6 +277,10 @@ def __init__(self, target: CMakeTarget, env: 'Environment', for_machine: Machine
             # Register the new languages and initialize the compile opts array
             for lang in languages:
                 self.languages.add(lang)
+
+                if self.clib_compiler is None:
+                    self.clib_compiler = self.env.coredata.compilers[self.for_machine].get(lang)
```

Comment:
I suspect that the requirement here is to select the broadest language as the best, so an order like:
`ObjC++, C++, ObjC, Fortran, C`? Although Fortran and cuda really make this more difficult. There is, very unfortunately, a function in compilers that does the opposite of what we want (and shoves non-clike languages to one end of the array).

I think something like:
```python
try:
   lang = sorted(languages, key=lambda x: ['cuda', 'objcpp', 'cpp', 'objc', 'fortran', 'c'].index(x))[0]
    self.clib_compiler = self.env.coredata.compilers[self.for_machine].get(lang)
except IndexError:
    pass
```

It would also be nice if we could add something to one of the cmake tests to cover this case.

Reply:
I don't think the order actually matters for "c-like" languages, does it? There's nothing stopping e.g. a C program from linking with an ObjC++ or Fortran library for example, and I believe the library search path should be shared as well… We're really using the compiler as a proxy for the linker here, which shouldn't care about the language.

I have absolutely no idea how this works for cuda though. I thought that thing only produces GPU kernels and not native executables, so using it to find host libraries makes no sense to me.

Reply:
The cuda runtime comes with libraries, which we may need to find. I'm fine with leaving cuda out for now, and worrying about that if/when it's actually a problem. In theory the linker search path could be different between them, but I guess that's more hypothetical than I can really come up with.

So, I'm convinced to not handle it at this level, but I would prefer if we set the clib_compiler outside the loop.

Reply:
> In theory the linker search path could be different between them, but I guess that's more hypothetical than I can really come up with.

I'm 95% sure that in that case everything is hopelessly broken anyway, because we don't know for which "compiler" these libraries are meant.

> So, I'm convinced to not handle it at this level, but I would prefer if we set the clib_compiler outside the loop.

Would something like this work, or do you still want a filter on it?

```python
try:
    self.clib_compiler = self.env.coredata.compilers[self.for_machine].get(languages[0])
except IndexError:
    pass
```

Reply:
Hm, small issue: languages is a set, which doesn't have an order and doesn't support indexing. Sure I could do `tuple(languages)[0]` to impose arbitrary order on it, but that doesn't seem… right? Just picking the first encountered language, as it currently does, or sorting it, as in your initial suggestion, seem like the more deterministic options. What do you think?